src/import/ImportWavPack.cpp

```diff
@@ -0,0 +1,298 @@
+{
+}
+
+wxString WavPackImportPlugin::GetPluginStringID()
+{
+   return wxT("libwavpack");
+}
+
+TranslatableString WavPackImportPlugin::GetPluginFormatDescription()
+{
+   return DESC;
+}
+
+std::unique_ptr<ImportFileHandle> WavPackImportPlugin::Open(const FilePath &filename, AudacityProject*)
+{
+   char errMessage[100]; // To hold possible error message
+   int flags = OPEN_WVC | OPEN_FILE_UTF8 | OPEN_TAGS;
+   WavpackContext *wavpackContext = WavpackOpenFileInput(filename, errMessage, flags, 0);
+   
+   if (!wavpackContext) {
+      // Some error occured(e.g. File not found or is invalid)
+      return nullptr;
+   }
+
+   auto handle = std::make_unique<WavPackImportFileHandle>(filename, wavpackContext);
+
+   return std::move(handle);
+}
+
+static Importer::RegisteredImportPlugin registered{ "WavPack",
+   std::make_unique< WavPackImportPlugin >()
+};
+
+// ============================================================================
+// WavPackImportFileHandle
+// ============================================================================
+
+WavPackImportFileHandle::WavPackImportFileHandle(const FilePath &filename,
+                                                WavpackContext *wavpackContext)
+:  ImportFileHandle(filename),
+   mWavPackContext(wavpackContext),
+   mNumChannels(WavpackGetNumChannels(mWavPackContext)),
+   mSampleRate(WavpackGetSampleRate(mWavPackContext)),
+   mBitsPerSample(WavpackGetBitsPerSample(mWavPackContext)),
+   mBytesPerSample(WavpackGetBytesPerSample(mWavPackContext)),
+   mNumSamples(WavpackGetNumSamples64(mWavPackContext))
+{
+   if (mBitsPerSample <= 16) {
+      mFormat = int16Sample;
+   } else if (mBitsPerSample <= 24) {
+      mFormat = int24Sample;
+   } else {
+      mFormat = floatSample;
+   }
+}
+
+TranslatableString WavPackImportFileHandle::GetFileDescription()
+{
+   return DESC;
+}
+
+auto WavPackImportFileHandle::GetFileUncompressedBytes() -> ByteCount
+{
+   return 0;
+}
+
+ProgressResult WavPackImportFileHandle::Import(WaveTrackFactory *trackFactory, TrackHolders &outTracks, Tags *tags)
+{
+   const int wavpackMode = WavpackGetMode(mWavPackContext);
+
+   outTracks.clear();
+
+   CreateProgress();
+
+   mChannels.resize(mNumChannels);
+
+   {
+      auto iter = mChannels.begin();
+      for (size_t c = 0; c < mNumChannels; ++iter, ++c)
+         *iter = NewWaveTrack(*trackFactory, mFormat, mSampleRate);
+   }
+
+   /* The number of samples to read in each loop */
+   const size_t SAMPLES_TO_READ = mChannels.begin()->get()->GetMaxBlockSize();
+   auto updateResult = ProgressResult::Success;
+   uint32_t totalSamplesRead = 0;
+
+   {
+      const uint32_t bufferSize = mNumChannels * SAMPLES_TO_READ;
+      ArrayOf<int32_t> wavpackBuffer{ bufferSize };
+      uint32_t samplesRead = 0;
+
+      do {
+         samplesRead = WavpackUnpackSamples(mWavPackContext, wavpackBuffer.get(), SAMPLES_TO_READ);
+
+         if (mFormat == int16Sample) {
+            ArrayOf<int16_t> tempBuffer{ bufferSize };
```

Comment:
Not that important here, but is is best to avoid memory allocations in such loops. Allocations can be quite costly.